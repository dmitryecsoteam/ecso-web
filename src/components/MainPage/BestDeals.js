import React from 'react';
import { Query } from 'react-apollo';
import { PulseLoader } from 'halogenium';

import TravelCard from './TravelCard';
import { GET_BEST_DEALS } from '../../queries/queries';

export default () => {

    return (
        <div className="best-deals__container">
            <h2 className="best-deals__text--big">Best Deals</h2>

            <Query query={GET_BEST_DEALS} variables={{ limit: 3, months: 3 }}>
                {({ data, loading, error }) => {

                    const spinner = <div className="results__spinner">
                        <PulseLoader color="#c1c1c1" />
                        <p className="results__text results__text--medium">Loading</p>
                    </div>

                    let travelsList = [];
                    if (data && data.getBestDeals) {
                        travelsList = data.getBestDeals.map((travel, i) => {
                            return <TravelCard
                                key={i}
                                _id={travel._id}
                                name={travel.destination.nameEn}
                                country={travel.destination.countryEn}
                                nameEn={travel.destination.nameEn}
                                countryEn={travel.destination.countryEn}
                                priceAirplane={travel.priceAirplane}
                                originName={travel.origin.nameEn}
                                date={travel.date}
                            />
                        })
                    }

                    return (
                        <div className="results__list">{loading ? spinner : travelsList}</div>
                    );

                }}

            </Query>


        </div>
    );

}