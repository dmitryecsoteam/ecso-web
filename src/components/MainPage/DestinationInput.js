import React from 'react';
import { startSearchDestinations } from '../../actions/destinationInputActions';
import { connect } from 'react-redux';
import selectDestinationSuggestions from '../../selectors/destinationInputSelector';
import Autosuggest from 'react-autosuggest';

const getSuggestionValue = (destination) => destination.name_en;

const renderSuggestion = (destinations) => (
    <span>
        {destinations.name_en}, {destinations.country_en}
    </span>
);

class DestinationInput extends React.Component {
    state = {
        enteredText: '',
        suggestDestinations: []
    };

    onChange = (event, { newValue }) => {

        // First, save previous enteredValue and update state with the new actual
        const prevValue = this.state.enteredText;
        this.setState({
            enteredText: newValue
        });

        // Second, check if it's the first letter in input and then fetch data from api
        if (this.state.enteredText === '') {
            this.props.startSearchDestinations(newValue).then(() => {
                this.setState({
                    suggestDestinations: selectDestinationSuggestions(this.props.destinations, this.state.enteredText)
                });
            });
        };
    };

    onSuggestionsFetchRequested = ({ value }) => {
        this.setState({
            suggestDestinations: selectDestinationSuggestions(this.props.destinations, value)
        });
    };

    onSuggestionsClearRequested = () => {
        this.setState({
            suggestDestinations: []
        });
    };

    onSuggestionSelected = (obj) => {
        console.log("From onSuggestionSelected: ", obj);
    };

    render() {
        const { enteredText, suggestDestinations } = this.state;

        const inputProps = {
            value: enteredText,
            onChange: this.onChange
        };

        return <div>
            <Autosuggest
                suggestions={suggestDestinations}
                onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
                onSuggestionsClearRequested={this.onSuggestionsClearRequested}
                getSuggestionValue={getSuggestionValue}
                renderSuggestion={renderSuggestion}
                inputProps={inputProps}
                onSuggestionSelected={this.onSuggestionSelected}
                focusInputOnSuggestionClick={false}
            />
        </div>

    };
};

const mapStateToProps = (state) => ({
    destinations: state.destinationInput.destinations,
    isFetching: state.destinationInput.isFetching
});

const mapDispatchToProps = (dispatch) => ({
    startSearchDestinations: (text) => dispatch(startSearchDestinations(text))
});

export default connect(mapStateToProps, mapDispatchToProps)(DestinationInput);