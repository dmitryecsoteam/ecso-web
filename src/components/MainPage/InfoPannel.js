import React, { useState } from 'react';

export default () => {

    const [isOpen, setIsOpen] = useState(true);

    console.log(isOpen)

    return (
        <React.Fragment>
            {isOpen && (
                <div className="info__container">
                    <button onClick={() => setIsOpen(false)} className="info__button">x</button>
                    <p className="info__text">Test data for origin: Tokyo, Osaka</p>
                </div>
                )}
        </React.Fragment>
    );

}

