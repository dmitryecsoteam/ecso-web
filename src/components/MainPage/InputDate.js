import React from 'react';
import { SingleDatePicker } from 'react-dates';
import isInclusivelyAfterDay from 'react-dates/src/utils/isInclusivelyAfterDay';
import isBeforeDay from 'react-dates/src/utils/isBeforeDay';
import moment from 'moment';


export default ({
    label,
    ...dateProps
}) => {

    const today = moment();
    const nextYear = moment().add(1, 'years');

    const isOutsideRange  = day => !(isInclusivelyAfterDay(day, today) && isBeforeDay(day, nextYear));

    return <div className="search-form__label-input">
        <label className="search-form__label">{label}</label>
        <SingleDatePicker
            {...dateProps}
            displayFormat={() => "DD/MM/YYYY"}
            numberOfMonths={1}
            isOutsideRange={isOutsideRange}
        />
    </div>
};