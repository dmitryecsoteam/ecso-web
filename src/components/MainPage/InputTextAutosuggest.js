import React from 'react';
import Autosuggest from 'react-autosuggest';
import city from '../../images/icons/city.png';

const getSuggestionValue = (suggestion) => (suggestion.nameEn);

const renderSuggestion = (suggestion) => (
    <div className="react-autosuggest__suggestion-content">
        <div className="react-autosuggest__suggestion-icon-container">
            <img className="react-autosuggest__suggestion-icon" src={city} />
        </div>
        <span>
            {suggestion.nameEn}, {suggestion.countryEn}
        </span>
    </div>
);

export default ({ label, error, errorText, ...props }) => {
    return <div className="search-form__autoinput">
        <label className="search-form__autoinput-label">{label}</label>
        <Autosuggest
            {...props}
            getSuggestionValue={getSuggestionValue}
            renderSuggestion={renderSuggestion}
        />
        {error && <span>{errorText}</span>}
    </div>
};