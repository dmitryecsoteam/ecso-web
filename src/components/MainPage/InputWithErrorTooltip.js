import React from 'react';
import classNames from 'classnames';

export default ({
    children,
    error = false,
    errorText = '',
    label = '',
    disabled = false
} = props) => {

    const divClassName = classNames(
        'search-form__label-input',
        { 'search-form__label-input--error': error },
        { 'search-form__label-input--disabled': disabled}
    );

    const errorClassName = classNames(
        'search-form__error',
        { 'search-form__error--open': error }
    );

    return (
        <div className="search-form__input-container">
            <div className={divClassName}>
                <label className="search-form__label">{label}</label>
                {children}
            </div>
            <div className={errorClassName}>
                <span className="search-form__error-text">{errorText}</span>
            </div>
        </div>
    );
}