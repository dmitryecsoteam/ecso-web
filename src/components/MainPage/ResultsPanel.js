import React from 'react';
import { connect } from 'react-redux';

import TravelCard from './TravelCard';

class ResultsPanel extends React.Component {

    shouldComponentUpdate(newProps) {
        // update results only if isFetching changed
        if (newProps.isFetching !== this.props.isFetching) return true;
        return false;
    }

    render() {

        const {
            isInitEmptyResult,
            travels,
            parametersPanel,
            parametersValue,
            isFetching
        } = this.props;



        let travelsList = travels.map((travel) => {

            const parameters = {};
            if (parametersPanel) {
                for (let param in parametersValue) {
                    if (parametersValue[param]) {
                        parameters[param] = travel.destination[param.toLowerCase() + 'Rating'];
                    }
                }
            }

            return (
                <TravelCard
                key={travel._id}
                _id={travel._id}
                name={travel.destination.nameEn}
                country={travel.destination.countryEn}
                priceAirplane={travel.priceAirplane}
                nameEn={travel.destination.nameEn}
                countryEn={travel.destination.countryEn}
                parameters={parameters}
            />
            );
        }
            
        );

        if (!isInitEmptyResult && travels.length === 0) {
            travelsList = <h3>Nothing was found</h3>;
        }

        return (
            <div className="results__list">
                {travelsList}
            </div>
        )
    };
};

const mapStatetoProps = (state) => ({
    travels: state.travels.result,
    isFetching: state.travels.isFetching,
    isInitEmptyResult: state.travels.isInitEmptyResult,
    parametersPanel: state.searchForm.parametersPanel,
    parametersValue: state.searchForm.parametersValue
});

export default connect(mapStatetoProps)(ResultsPanel);