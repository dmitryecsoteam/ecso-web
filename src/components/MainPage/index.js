import React from 'react';
import Header from '../Header';
import SearchForm from './SearchForm';
import ResultsPanel from './ResultsPanel';
import BestDeals from './BestDeals';



export default () => (
    <div>
        <div className="main-header">
            <Header />
            <div className="main-header__background">
                <img className="main-header__background-image" src="/images/background/background.jpg" alt="Background image Mount Fuji" />
            </div>
        </div>

        <SearchForm />

        <ResultsPanel />

        <BestDeals />
    </div>
);
