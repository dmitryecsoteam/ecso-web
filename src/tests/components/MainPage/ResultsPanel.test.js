import React from 'react';
import { shallow } from 'enzyme';
import { ResultsPanel } from '../../../components/MainPage/ResultsPanel';
import travels from '../../fixtures/travels';

let wrapper, setFilter;
beforeEach(() => {
    setFilter = jest.fn();
    wrapper = shallow(<ResultsPanel
        travels={[]}
        isFetching={false}
        isInitEmptyResult={true}
        parametersPanel={false}
        parametersValue={
            {
                Beach: 0,
                Food: 0,
                Museum: 0,
                Nature: 0,
                Shopping: 0,
                Nightlife: 0
            }
        }
        filter={
            {
                sortBy: 'relevance',
                desc: true
            }
        }
        setFilter={setFilter}
    />);
});

test('should render emplty div on INIT', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render spinner', () => {
    wrapper.setProps({ isFetching: true });
    expect(wrapper).toMatchSnapshot();
});

test('should render single TravelCard without filter', () => {
    wrapper.setProps({ isFetching: true });
    wrapper.setProps({
        travels: [travels[0]],
        isInitEmptyResult: false,
        isFetching: false
    });
    expect(wrapper).toMatchSnapshot();
});

test('should render list of travels', () => {
    wrapper.setProps({ isFetching: true });
    wrapper.setProps({
        travels: [travels[0], travels[1]],
        isInitEmptyResult: false,
        isFetching: false,
        parametersPanel: true,
        parametersValue: {
            Beach: 0,
            Food: 4,
            Museum: 4,
            Nature: 0,
            Shopping: 0,
            Nightlife: 0
        }
    });
    expect(wrapper).toMatchSnapshot();
});

test('should render list of travels with price descending filter', () => {
    wrapper.setProps({ isFetching: true });
    wrapper.setProps({
        travels: [travels[0], travels[1]],
        isInitEmptyResult: false,
        isFetching: false,
        parametersPanel: true,
        parametersValue: {
            Beach: 0,
            Food: 4,
            Museum: 4,
            Nature: 0,
            Shopping: 0,
            Nightlife: 0
        },
        filter: {
            sortBy: 'price',
            desc: true
        }
    });
    expect(wrapper).toMatchSnapshot();
});

test('should set filters on click', () => {
    wrapper.setProps({ isFetching: true });
    wrapper.setProps({
        travels: [travels[0], travels[1]],
        isInitEmptyResult: false,
        isFetching: false,
        parametersPanel: true,
        parametersValue: {
            Beach: 0,
            Food: 4,
            Museum: 4,
            Nature: 0,
            Shopping: 0,
            Nightlife: 0
        }
    });

    // shouldn't set current filter
    wrapper.find('.results__filter').simulate('click', { target: { id: 'relevanceDesc' } });
    expect(setFilter).toHaveBeenCalledTimes(0);

    wrapper.setProps({ filter: { sortBy: 'relevance', desc: false } });
    wrapper.find('.results__filter').simulate('click', { target: { id: 'relevanceAsc' } });
    expect(setFilter).toHaveBeenCalledTimes(0);

    wrapper.setProps({ filter: { sortBy: 'price', desc: true } });
    wrapper.find('.results__filter').simulate('click', { target: { id: 'priceDesc' } });
    expect(setFilter).toHaveBeenCalledTimes(0);

    wrapper.setProps({ filter: { sortBy: 'price', desc: false } });
    wrapper.find('.results__filter').simulate('click', { target: { id: 'priceAsc' } });
    expect(setFilter).toHaveBeenCalledTimes(0);

    wrapper.setProps({ filter: { sortBy: 'relevance', desc: true } });


    // should set filter action
    wrapper.find('.results__filter').simulate('click', { target: { id: 'relevanceAsc' } });
    expect(setFilter).toHaveBeenCalledTimes(1);
    expect(setFilter).toHaveBeenCalledWith('relevance', false);

    wrapper.find('.results__filter').simulate('click', { target: { id: 'priceDesc' } });
    expect(setFilter).toHaveBeenCalledTimes(2);
    expect(setFilter).toHaveBeenCalledWith('price', true);

    wrapper.find('.results__filter').simulate('click', { target: { id: 'priceAsc' } });
    expect(setFilter).toHaveBeenCalledTimes(3);
    expect(setFilter).toHaveBeenCalledWith('price', false);
});