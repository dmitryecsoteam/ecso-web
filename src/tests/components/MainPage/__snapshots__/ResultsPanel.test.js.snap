// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render emplty div on INIT 1`] = `
<div
  className="results__list"
/>
`;

exports[`should render list of travels 1`] = `
<div
  className="results__list"
>
  <p
    className="results__filter"
    onClick={[Function]}
  >
    sort by: 
    <span
      className="results__filter-name--active"
    >
      relevance
    </span>
    <span
      className="results__filter-arrow results__filter-arrow--active"
      id="relevanceDesc"
    >
      ▼
    </span>
    <span
      className="results__filter-arrow"
      id="relevanceAsc"
    >
      ▲
    </span>
      
    <span
      className=""
    >
      price
    </span>
    <span
      className="results__filter-arrow"
      id="priceDesc"
    >
      ▼
    </span>
    <span
      className="results__filter-arrow"
      id="priceAsc"
    >
      ▲
    </span>
  </p>
  <TravelCard
    _id="5c6c7e42e3b7e1edb974a39q"
    carDistance={150}
    country="Japan"
    countryEn="Japan"
    key="5c6c7e42e3b7e1edb974a39q"
    name="Nara"
    nameEn="Nara"
    parameters={
      Object {
        "Food": 4,
        "Museum": 5,
      }
    }
    priceAirplane={100}
  />
  <TravelCard
    _id="5c6c7e42e3b7e1edb974a39f"
    carDistance={512}
    country="Japan"
    countryEn="Japan"
    key="5c6c7e42e3b7e1edb974a39f"
    name="Osaka"
    nameEn="Osaka"
    parameters={
      Object {
        "Food": 5,
        "Museum": 4,
      }
    }
    priceAirplane={115}
  />
  <div
    className="results__divider"
  />
</div>
`;

exports[`should render list of travels with price descending filter 1`] = `
<div
  className="results__list"
>
  <p
    className="results__filter"
    onClick={[Function]}
  >
    sort by: 
    <span
      className=""
    >
      relevance
    </span>
    <span
      className="results__filter-arrow"
      id="relevanceDesc"
    >
      ▼
    </span>
    <span
      className="results__filter-arrow"
      id="relevanceAsc"
    >
      ▲
    </span>
      
    <span
      className="results__filter-name--active"
    >
      price
    </span>
    <span
      className="results__filter-arrow results__filter-arrow--active"
      id="priceDesc"
    >
      ▼
    </span>
    <span
      className="results__filter-arrow"
      id="priceAsc"
    >
      ▲
    </span>
  </p>
  <TravelCard
    _id="5c6c7e42e3b7e1edb974a39f"
    carDistance={512}
    country="Japan"
    countryEn="Japan"
    key="5c6c7e42e3b7e1edb974a39f"
    name="Osaka"
    nameEn="Osaka"
    parameters={
      Object {
        "Food": 5,
        "Museum": 4,
      }
    }
    priceAirplane={115}
  />
  <TravelCard
    _id="5c6c7e42e3b7e1edb974a39q"
    carDistance={150}
    country="Japan"
    countryEn="Japan"
    key="5c6c7e42e3b7e1edb974a39q"
    name="Nara"
    nameEn="Nara"
    parameters={
      Object {
        "Food": 4,
        "Museum": 5,
      }
    }
    priceAirplane={100}
  />
  <div
    className="results__divider"
  />
</div>
`;

exports[`should render single TravelCard without filter 1`] = `
<div
  className="results__list"
>
  <TravelCard
    _id="5c6c7e42e3b7e1edb974a39f"
    carDistance={512}
    country="Japan"
    countryEn="Japan"
    key="5c6c7e42e3b7e1edb974a39f"
    name="Osaka"
    nameEn="Osaka"
    parameters={Object {}}
    priceAirplane={115}
  />
  <div
    className="results__divider"
  />
</div>
`;

exports[`should render spinner 1`] = `
<div
  className="results__list"
>
  <div
    className="results__spinner"
  >
    <PulseLoader
      color="#c1c1c1"
      loading={true}
      margin="2px"
      size="15px"
    />
    <p
      className="results__text results__text--medium"
    >
      Searching
    </p>
  </div>
  <div
    className="results__divider"
  />
</div>
`;
